File: http.ts
import {rawlog, log} from "./util";

// import * as util from "./util";
import { pubInternal, sub } from "./dispatch";
import { main as pb } from "./msg.pb";

const servers = new Map<number, Server>();
const dec = new TextDecoder("utf8");
const enc = new TextEncoder();

export function initHttp() {
  sub("http", (payload: Uint8Array) => {
      const msg = pb.Msg.decode(payload);
      const id = msg.httpServerId;
      const s = servers.get(id);
      s.onMsg(msg);
  });
}

export function abc() {
    rawlog("abc is called!");
    return 1+1;
}

export interface RequestOptions {
    method?: string;
    referrer?: string;
    mode?: string;
    credentials?: string;
    redirect?: string;
    integrity?: string;
    cache?: string;
}

export class Request {
    opts: RequestOptions;
    readonly default_method = "GET";
    path: string;
    method: string;
    body: any | string;
    constructor(url: string, opts?: RequestOptions) {
      if(opts == null) {
        opts = {
            method: this.default_method
        }
      }
      this.opts = opts;
      this.method = opts.method;
    }
}

export class Response {
    server_id: number;
    req_id: number;
    constructor(Server?: Server) {
    }
    write(data: Uint8Array | string) {
        let raw_data: Uint8Array
        if (typeof(data) == "string") {
            raw_data = enc.encode(data);
        } else {
            raw_data = data as Uint8Array;
        }
        pubInternal("http", {
            command: pb.Msg.Command.HTTP_RES_WRITE,
            // httpServerId: this.server_id,
            httpReqId: this.req_id,
            httpResBody: raw_data
        })
    }
    status(code: number) {
        pubInternal("http", {
            command: pb.Msg.Command.HTTP_RES_STATUS,
            httpReqId: this.req_id,
            httpResCode: code
        })
    }
    end() {
        pubInternal("http", {
            command: pb.Msg.Command.HTTP_RES_END,
            httpServerId: this.server_id,
            httpReqId: this.req_id
        })
    }
}

let nextServerId = 0;
export class Server {
    private readonly id: number;
    private port: number;
    private cb: (req: Request, res: Response) => void;
    constructor(cb: (req: Request, res: Response) => void) {
        this.id = nextServerId++;
        this.cb = cb;
        servers.set(this.id, this);
        pubInternal("http", {
            command: pb.Msg.Command.HTTP_CREATE,
            httpServerId: this.id
        })
    }
    listen(port: number) {
        log("Starting server on", port);
        pubInternal("http", {
            command: pb.Msg.Command.HTTP_LISTEN,
            httpServerId: this.port,
            httpListenPort: port
        })
    }
    buildRequest(msg: pb.Msg) {
        let req = new Request("", {});
        req.path = msg.httpReqPath;
        req.method = msg.httpReqMethod;
        let raw_body = dec.decode(msg.httpReqBody);
        // Implement a JSON body parser by default, fallback to string representation:
        try {
            let body = JSON.parse(raw_body)
            req.body = body;
        } catch(e) {
            req.body = raw_body.toString();
        };
        return req
    }
    buildResponse(msg: pb.Msg) {
        let res = new Response();
        res.server_id = msg.httpServerId;
        res.req_id = msg.httpReqId;
        return res;
    }
    onMsg(msg: pb.Msg) {
        if (msg.command = pb.Msg.Command.HTTP_REQ) {
            let req = this.buildRequest(msg);
            let res = this.buildResponse(msg);
            this.cb(req, res);
        }
    }
}

export function createServer(cb: (req: Request, res: Response)=>void): Server {
    let s = new Server(cb);
    return s;
}

File: http.go
// Copyright 2018 Ryan Dahl <ry@tinyclouds.org>
// All rights reserved. MIT License.
package deno

import (
	"fmt"
	"io/ioutil"
	"net/http"
	"sync"
	"sync/atomic"

	"github.com/golang/protobuf/proto"
)

const (
	httpChan     = "http"
	serverHeader = "Deno"
)

var (
	requestDispatcher *RequestDispatcher
	httpServers       = make(map[int32]*http.Server)
)

// RequestDispatcher takes care of all request-response related commands.
type RequestDispatcher struct {
	// reqs  map[int32]*Request
	reqs  *sync.Map
	reqID *int32
}

// Store constructs an internal request object and stores it into the requests map, returns a pointer to it.
func (d *RequestDispatcher) Store(req *http.Request, res http.ResponseWriter) *Request {
	atomic.AddInt32(d.reqID, 1)
	id := *d.reqID
	r := &Request{
		req:  req,
		res:  res,
		done: make(chan bool),
		id:   id,
	}
	d.reqs.Store(id, r)
	return r
}

// Dispatch will take care of all HTTP_RES_* commands.
func (d *RequestDispatcher) Dispatch(msg *Msg) bool {
	reqPtr, ok := d.reqs.Load(msg.HttpReqId)
	// TODO: handle non-ok here:
	if !ok {
		return ok
	}
	req := reqPtr.(*Request)
	switch msg.Command {
	case Msg_HTTP_RES_WRITE:
		req.res.Write(msg.HttpResBody)
	case Msg_HTTP_RES_STATUS:
		req.res.WriteHeader(int(msg.HttpResCode))
	case Msg_HTTP_RES_END:
		go func() {
			req.done <- true
		}()
	}
	return true
}

// Request wraps http.Request and http.ResponseWriter with additional helper methods.
type Request struct {
	req *http.Request
	res http.ResponseWriter
	id  int32

	// This is used to signal "req.end()":
	done chan bool
}

// ToMsg is a helper method that builds up the request PB object.
func (r *Request) ToMsg() (msg *Msg) {
	msg = &Msg{
		Command: Msg_HTTP_REQ,
		// HttpServerId:  serverID,
		HttpReqPath:   r.req.URL.Path,
		HttpReqMethod: r.req.Method,
		HttpReqId:     r.id,
	}
	return msg
}

func InitHTTP() {
	// Initialize the request dispatcher:
	// var n int32 = 1
	requestDispatcher = &RequestDispatcher{
		reqs:  &sync.Map{},
		reqID: new(int32),
	}
	Sub(httpChan, func(buf []byte) []byte {
		msg := &Msg{}
		check(proto.Unmarshal(buf, msg))
		// If the request ID is present use the request dispatcher:
		if msg.HttpReqId > 0 {
			requestDispatcher.Dispatch(msg)
			return buf
		}

		switch msg.Command {
		case Msg_HTTP_CREATE:
			httpServers[msg.HttpServerId] = &http.Server{}
		case Msg_HTTP_LISTEN:
			httpListen(msg.HttpServerId, msg.HttpListenPort)
		default:
			panic("[http] Unexpected message " + string(buf))
		}
		return buf
	})
}

func buildHTTPHandler(serverID int32) func(w http.ResponseWriter, r *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		req := requestDispatcher.Store(r, w)
		var body []byte
		if r.Body != nil {
			body, _ = ioutil.ReadAll(r.Body)
		}
		msg := req.ToMsg()
		msg.HttpReqBody = body
		go PubMsg(httpChan, msg)

		w.Header().Set("Server", serverHeader)

		// Wait for end signal:
		<-req.done
	}
}

func httpListen(serverID int32, port int32) {
	s := httpServers[serverID]
	listenAddr := fmt.Sprintf(":%d", port)
	handler := buildHTTPHandler(serverID)
	s.Addr = listenAddr
	s.Handler = http.HandlerFunc(handler)
	go s.ListenAndServe()

}
